temporalio/bridge/worker.py

```diff
@@ -136,3 +154,199 @@ def to_bridge_payload(
     return temporalio.bridge.proto.common.Payload(
         metadata=payload.metadata, data=payload.data
     )
+
+
+# See https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-classes-that-are-generic-in-stubs-but-not-at-runtime
+if TYPE_CHECKING:
+    BridgePayloadContainer = (
+        google.protobuf.internal.containers.RepeatedCompositeFieldContainer[
+            temporalio.bridge.proto.common.Payload
+        ]
+    )
+else:
+    BridgePayloadContainer = (
+        google.protobuf.internal.containers.RepeatedCompositeFieldContainer
+    )
+
+
+async def _apply_to_bridge_payloads(
+    payloads: BridgePayloadContainer,
+    cb: Callable[
+        [Iterable[temporalio.api.common.v1.Payload]],
+        Awaitable[List[temporalio.api.common.v1.Payload]],
+    ],
+) -> None:
+    if len(payloads) == 0:
+        return
+    new_payloads = to_bridge_payloads(await cb(from_bridge_payloads(payloads)))
+    del payloads[:]
+    # TODO(cretz): Copy too expensive?
+    payloads.extend(new_payloads)
+
+
+async def _apply_to_bridge_payload(
+    payload: temporalio.bridge.proto.common.Payload,
+    cb: Callable[
+        [Iterable[temporalio.api.common.v1.Payload]],
+        Awaitable[List[temporalio.api.common.v1.Payload]],
+    ],
+) -> None:
+    new_payload = (await cb([from_bridge_payload(payload)]))[0]
+    payload.metadata.clear()
+    payload.metadata.update(new_payload.metadata)
+    payload.data = new_payload.data
+
+
+async def _decode_bridge_payloads(
+    payloads: BridgePayloadContainer,
+    codec: temporalio.converter.PayloadCodec,
+) -> None:
+    return await _apply_to_bridge_payloads(payloads, codec.decode)
+
+
+async def _decode_bridge_payload(
+    payload: temporalio.bridge.proto.common.Payload,
+    codec: temporalio.converter.PayloadCodec,
+) -> None:
+    return await _apply_to_bridge_payload(payload, codec.decode)
+
+
+async def _encode_bridge_payloads(
+    payloads: BridgePayloadContainer,
+    codec: temporalio.converter.PayloadCodec,
+) -> None:
+    return await _apply_to_bridge_payloads(payloads, codec.encode)
+
+
+async def _encode_bridge_payload(
+    payload: temporalio.bridge.proto.common.Payload,
+    codec: temporalio.converter.PayloadCodec,
+) -> None:
+    return await _apply_to_bridge_payload(payload, codec.encode)
+
+
+async def decode_activation(
```

Comment:
In TS we managed to have the type checker verify we encode and decode all relevant fields.
Is this possible in Python? If not, is there some other way we can verify it?

Reply:
No, in TS we mark Payloads so the type checker complains if something is missed.
I relied on this when we added new fields to the proto.

https://github.com/temporalio/sdk-typescript/blob/main/packages/internal-non-workflow-common/src/codec-types.ts

Reply:
I see. I am afraid the type system is too weak to do any such thing. I can probably come up with a test where I walk the proto descriptor and set every possible payload (one at a time due to oneof and other restrictions) and confirm it was encoded (save for search attributes). It's non-trivial, but maybe not _too_ hard. Think it's worth it?